firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +35,115 @@
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    // Android provides lifecycle callbacks (eg onStop()) that custom `Activity`s can extend. But we
+    // can't take advantage of that, since we need to be usable with a generic Activity. So instead,
+    // we create a custom Fragment, and add that Fragment to the given Activity. When the Activity
+    // stops, it will automatically stop the attached Fragments too.
+    //
+    // One difficulty with this approach is that how you get a Fragment and attach it to an Activity
+    // differs based on the type of Activity. If the Activity is actually a FragmentActivity, then
+    // you must use the android.support.v4.app.FragmentManager to do so. Otherwise, you need to use
+    // the deprecated android.app.FragmentManager.
+    //
+    // Possible improvements:
+    // 1) Allow other lifecycle callbacks other than just 'onStop'.
+    // 2) Use LifecycleOwner (which FragmentActivity implements, but Activity does not) to register
+    //    for lifecycle callbacks instead of creating/attaching a Fragment.
+
+    if (activity instanceof FragmentActivity) {
+      FragmentActivity fragmentActivity = (FragmentActivity) activity;
+
+      StopListenerSupportFragment fragment = null;
+      try {
+        fragment =
+            (StopListenerSupportFragment)
+                fragmentActivity
+                    .getSupportFragmentManager()
+                    .findFragmentByTag(SUPPORT_FRAGMENT_TAG);
+      } catch (ClassCastException e) {
+        throw new IllegalStateException(
+            "Fragment with tag '" + SUPPORT_FRAGMENT_TAG + "' is not a StopListenerSupportFragment",
+            e);
+      }
+
+      if (fragment == null || fragment.isRemoving()) {
+        fragment = new StopListenerSupportFragment();
+        fragmentActivity
+            .getSupportFragmentManager()
+            .beginTransaction()
+            .add(fragment, SUPPORT_FRAGMENT_TAG)
+            .commitAllowingStateLoss();
+      }
+
+      fragment.addCallback(callback);
+
+    } else {
+      // If we get here, then we have a non-FragmentActivity Activity. Unfortunately, all Fragment
+      // related classes/methods with non-FragmentActivity Activity's are deprecated, implying that
+      // almost everything in this block is deprecated.
+
+      StopListenerFragment fragment = null;
+      try {
+        fragment =
+            (StopListenerFragment) activity.getFragmentManager().findFragmentByTag(FRAGMENT_TAG);
+      } catch (ClassCastException e) {
+        throw new IllegalStateException(
+            "Fragment with tag '" + FRAGMENT_TAG + "' is not a StopListenerFragment", e);
+      }
+
+      if (fragment == null || fragment.isRemoving()) {
```

Comment:
Can instances ListenerRegistrationImpl be created concurrently from different threads? If so there might be a data race as the following steps are not atomic:
1. you look for a fragment
2. it's not there
3. you insert one

Please double check if you need to synchronize, here and above.

Reply:
Good catch. Most of firestore gets synchronized via our async queue, but I don't think this does. I've put this in a sync block.

But it gets worse: FragmentManager::commit*() doesn't actually... commit. Instead, it schedules the action for sometime later. So a synch block is necessary, but not sufficient. I've added a cache map from activity to fragment and now check the cache first.

Reply:
> FragmentManager::commit*() doesn't actually... commit

hm, it seems like using a fragment cache to overcome that adds a non-trivial amount of complexity like potential memory leaks, dangling references in weakptrs, etc.

Have you considered running all the registration logic on the [UI thread](https://developer.android.com/training/multiple-threads/communicate-ui)? That would remove the need to synchronize as all registrations will happen sequentially, and coupled with using `commitNow()`(which I *think* must be called on the UI thread as well) instead of `commit()` should fix the problem, wdyt?